cmd/localstack/awsutil.go

```diff
@@ -196,6 +201,178 @@ func DownloadCodeArchive(url string) {
+		watcher:       watcher,
+	}, nil
+}
+
+func (c *ChangeListener) watch() {
+	for {
+		select {
+		case event, ok := <-c.watcher.Events():
+			if !ok {
+				return
+			}
+			log.Debugln("FileWatcher got event: ", event)
+			if event.Has(fsnotify.Create) {
+				stat, err := os.Stat(event.Name)
+				if err != nil {
+					log.Errorln("Error stating event file: ", event.Name, err)
+				} else if stat.IsDir() {
+					subfolders := getSubFolders(event.Name)
+					for _, folder := range subfolders {
+						err = c.watcher.Add(folder)
+						c.watchedFolders = append(c.watchedFolders, folder)
+						if err != nil {
+							log.Errorln("Error watching folder: ", folder, err)
+						}
+					}
+				}
+				// remove in case of remove / rename (rename within the folder will trigger a separate create event)
+			} else if event.Has(fsnotify.Remove) || event.Has(fsnotify.Rename) {
+				// remove all file watchers if it is in our folders list
+				toBeRemovedDirs, newWatchedFolders := getSubFoldersInList(event.Name, c.watchedFolders)
+				c.watchedFolders = newWatchedFolders
+				for _, dir := range toBeRemovedDirs {
+					err := c.watcher.Remove(dir)
+					if err != nil {
+						log.Warnln("Error removing path: ", event.Name, err)
+					}
+				}
+			}
+			c.changeChannel <- event.Name
+		case err, ok := <-c.watcher.Errors():
+			if !ok {
+				log.Println("error:", err)
+				return
+			}
+			log.Println("error:", err)
+		}
+	}
+}
+
+func (c *ChangeListener) addTargetPaths(targetPaths []string) {
+	// Add all target paths and subfolders
+	for _, targetPath := range targetPaths {
+		subfolders := getSubFolders(targetPath)
+		log.Infoln("Subfolders: ", subfolders)
+		for _, target := range subfolders {
+			err := c.watcher.Add(target)
+			if err != nil {
+				log.Fatal(err)
+			}
+		}
+	}
+}
+
+func (c *ChangeListener) close() error {
+	return c.watcher.Close()
+}
+
+func resetListener(changeChannel <-chan bool, server *CustomInteropServer) {
+	for {
+		_, more := <-changeChannel
+		if !more {
+			return
+		}
+		log.Println("Resetting environment...")
+		_, err := server.Reset("HotReload", 2000)
+		if err != nil {
+			log.Warnln("Error resetting server: ", err)
+		}
+	}
+
+}
+
+func debounceChannel(changeChannel <-chan string, duration time.Duration) <-chan bool {
+	resultChannel := make(chan bool, 10)
+	// debouncer to limit restarts
+	timer := time.NewTimer(duration)
+	// immediately stop the timer, since we do not want to reload right at the startup
+	if !timer.Stop() {
+		// we have to drain the channel in case the timer already fired
+		<-timer.C
+	}
+	go func() {
+		for {
+			select {
+			case _, more := <-changeChannel:
+				if !more {
+					timer.Stop()
+					close(resultChannel)
+					return
+				}
+				timer.Reset(duration)
+			case <-timer.C:
+				resultChannel <- true
+			}
+		}
+	}()
+	return resultChannel
+}
+
+func RunHotReloadingListener(server *CustomInteropServer, targetPaths []string, opts *LsOpts, ctx context.Context) {
```

Comment:
suggestion: separate this more from the ChangeListener and export the ChangeListener "API"